README.md
# Photon Machine Learning (Photon ML)

[![Build Status](https://travis-ci.org/linkedin/photon-ml.svg?branch=master)](https://travis-ci.org/linkedin/photon-ml)

**Check out our [hands-on tutorial](https://github.com/linkedin/photon-ml/wiki/Photon-ML-Tutorial).**

Photon ML is a machine learning library based on Apache Spark. It was originally developed by the LinkedIn Machine Learning Algorithms Team. Currently, Photon ML supports training different types of [Generalized Linear Models](https://en.wikipedia.org/wiki/Generalized_linear_model)(GLMs) and [Generalized Linear Mixed Models](https://en.wikipedia.org/wiki/Generalized_linear_mixed_model)(GLMMs/GLMix model): logistic, linear, and Poisson.

- [Features](#features)
  - [Generalized Linear Models](#generalized-linear-models)
  - [GAME - Generalized Additive Mixed Effects](#game---generalized-additive-mixed-effects)
  - [Configurable Optimizers](#configurable-optimizers)
  - [Regularization](#regularization)
  - [Feature scaling and normalization](#feature-scaling-and-normalization)
  - [Offset training](#offset-training)
  - [Feature summarization](#feature-summarization)
  - [Model validation](#model-validation)
  - [Warm-start training](#warm-start-training)
  - [Partial re-training](#partial-re-training)
- [Experimental Features](#experimental-features)
  - [Smoothed Hinge Loss Linear SVM](#smoothed-hinge-loss-linear-svm)
  - [Hyperparameter Auto-Tuning](#hyperparameter-auto-tuning)
  - [Regularize by Previous Model During Warm-Start Training](#regularize-by-previous-model-during-warm-start-training)
- [How to Build](#how-to-build)
- [How to Use](#how-to-use)
  - [Drivers](#drivers)
  - [API](#api)
  - [Avro Schemas](#avro-schemas)
  - [What about other formats?](#what-about-other-formats)
  - [Input Data Format](#input-data-format)
  - [Models](#models)
  - [Shaded Jar](#shaded-jar)
  - [Try It Out!](#try-it-out)
    - [Install Spark](#install-spark)
    - [Get and Build the Code](#get-and-build-the-code)
    - [Grab a Dataset](#grab-a-dataset)
    - [Train the Model](#train-the-model)
  - [Running Photon ML on Cluster Mode](#running-photon-ml-on-cluster-mode)
- [Modules and directories](#modules-and-directories)
  - [Source code](#source-code)
  - [Other](#other)
- [IntelliJ IDEA setup](#intellij-idea-setup)
- [How to Contribute](#how-to-contribute)
- [Reference](#reference)

## Features

#### Generalized Linear Models

  * Linear Regression
  * Logistic Regression
  * Poisson Regression
  
#### GAME - Generalized Additive Mixed Effects

The GAME algorithm uses coordinate descent to expand beyond traditional GLMs to further provide per-entity (per-user, per-item, per-country, etc.) coefficients (also known as random effects in statistics literature). It manages to scale model training up to hundreds of billions of coefficients, while remaining solvable within Spark's framework.

For example, a GAME model for movie recommendations can be formulated as (fixed effect model + per-user random effect model + per-movie random effect model + user-movie matrix factorization model). More details on GAME models can be found [here](https://docs.google.com/presentation/d/1vHanpK3KLIVgdDIHYRehUeyb04Hc2AasbBHs4InVPSU).

The type of GAME model currently supported by Photon ML is the GLMM or 'GLMix' model. Many of LinkedIn's core products have adopted GLMix models: jobs search and recommendation, news feed ranking, Ads CTR prediction and "People Also Viewed". More details on GLMix models can be found [here](https://docs.google.com/presentation/d/1tYoelUma9-MMYdteWYS31LqVeoyPEncxJRk-k57gj0A/edit?usp=sharing).

#### Configurable Optimizers
  * [LBFGS](https://en.wikipedia.org/wiki/Limited-memory_BFGS)
  * [TRON](https://www.csie.ntu.edu.tw/~cjlin/papers/logistic.pdf)

#### Regularization
  * L1 (LASSO) regularization
  * L2 (Tikhonov) regularization (only type supported by TRON)
  * Elastic-net regularization

#### Feature scaling and normalization
  * Standardization: Zero-mean, unit-variant normalization
  * Scaling by standard deviation
  * Scaling by maximum magnitude to range [-1, 1]

#### Offset training
A typical naive way of training multi-layer models, it's used to insert another model's response into a global model. For example, when doing a typical binary classification problem, a model could be trained against a subset of all the features. Next, data could be scored with this model and the response scores set as 'offset' values. In this way, future models will learn against the residuals of the 1st layer model's response while having the benefits of combining the two models together.

#### Feature summarization
Provides typical metrics (mean, min, max, std, variance, etc.) on a per feature basis.
  
#### Model validation
Compute evaluation metrics for the trained models over a validation dataset, such as AUC, RMSE, or Precision@k.
  
#### Warm-start training
Load existing models and use their coefficients as a starting point for optimization. When training multiple models in succession, use the coefficients of the previous model.
  
#### Partial re-training
Load existing models, but lock their coefficients. Allows efficient re-training of portions of a GAME model.

## Experimental Features
Photon ML currently contains a number of experimental features that have not been fully tested.

#### Smoothed Hinge Loss Linear SVM
In addition to the Generalized Linear Models described above, Photon-ML also supports an optimizer-friendly approximation for linear SVMs as described [here](http://qwone.com/~jason/writing/smoothHinge.pdf) by Jason D. M. Rennie.

#### Hyperparameter Auto-Tuning
Automatically explore the hyperparameter space for your GAME model. Two types of search exist:
- Random search: Use Sobol sequences to randomly, but evenly, explore the hyperparameter space
- Bayesian search: Use a Gaussian process to perform a directed search throughout the hyperparameter space

#### Regularize by Previous Model During Warm-Start Training
Use the means and variances of an existing model to help approximate good coefficients when doing warm-start training. Allows multiple rounds of warm-start training on small datasets to achieve parity with cold-start training on a single large dataset.

## How to Build
**Note**: Before building, please make sure environment variable ```JAVA_HOME``` is pointed at a Java 8 JDK property. Photon ML is not compatible with JDK < 1.8.
The below commands are for Linux/Mac users, for Windows, please use ```gradlew.bat``` instead of ```gradlew```.

```bash
# Build binary jars only:
./gradlew assemble

# Build with all tests (unit and integration):
./gradlew clean build

# Build with only unit tests:
./gradlew clean build -x integTest

# Build with only integration tests:
./gradlew clean build -x test

# Build with no tests:
./gradlew clean build -x test -x integTest

# Run unit tests:
./gradlew test

# Run integration tests:
./gradlew integTest

# Check License with Apache Rat
./gradlew rat

# Check scala style
./gradlew scalastyle

# Check everything
./gradlew check
```
## How to Use

### Drivers
To use Photon ML from the command line, 3 default drivers exist: the Legacy Photon driver for GLM training, the GAME training driver, and the GAME scoring driver. Each of these have their own input parameters. We recommend using the GAME drivers, as a GLM is a special case of GAME model. The Legacy Photon driver has not been developed for some time and is deprecated.

### API
Photon ML can be imported just like Spark ML, and the API layer used directly. Where possible, we have tried to make the interfaces identical to those of Spark ML. See the driver source code for examples of how to use the Photon ML API.

### Avro Schemas
The currently available drivers read/write data in [Apache Avro](https://avro.apache.org/) format. The detailed schemas are declared at [photon-avro-schemas](https://github.com/linkedin/photon-ml/tree/master/photon-avro-schemas/src/main/avro) module.

### What about other formats?
LinkedIn uses primarily Avro formatted data. While Avro does provide a unified and rigorous way of managing all critical data representations, we think it is also important to allow other data formats to make Photon ML more flexible. Contributions of DataReaders for other formats to Photon ML are welcome and encouraged.

### Input Data Format

Photon ML reserves the following field names in the Avro input data:
1. **response**: `double` (required)
    - The response/label for the event
2. **weight**: `double` (optional)
    - The relative weight of a particular sample compared to other samples
    - Default = 1.0
3. **offset**: `double` (optional)
    - The residual score computed by some other model
    - Default = 0.0
    - Computed scores always take the form `(x * B) + offset`, where `x` is the feature vector and `B` is the coefficient vector
4. **uid**: `string`, `int`, or `long` (optional)
    - A unique ID for the sample
5. **metadataMap**: `map: [string]` (optional)
    - A map of non-feature metadata for the sample
6. **features**: `array: [FeatureAvro]` (required by Legacy Photon driver)
    - An array of features to use for training/scoring

All of these default names can be overwritten using the GAME drivers. However, they are reserved and cannot be used for purposes other than their default usage (e.g. cannot specify `response` as your weight column).

Additional fields may exist in the record, and in fact are necessary for certain features (e.g. must have ID fields to group data by for random effect models or certain validation metrics).

Features loaded through the existing drivers are expected to follow the LinkedIn naming convention. Each feature must be an Avro record with the following fields:
1. **name**: `string`
  - The feature name/category
2. **term**: `string`
  - The feature sub-category
3. **value**: `double`
  - The feature value
  
To demonstrate the difference between `name` and `term`, consider the following categorical features:
```
  name = "age"
  term = "0-10"
  value = 1.0
  
  name = "age"
  term = "11-20"
  value = 0.0
  
  ...
```

### Models

Legacy Photon outputs model coefficients directly to text:
```bash
# For each line in the text file:
[feature_string]\t[feature_id]\t[coefficient_value]\t[regularization_weight]
```

GAME models are output using the [BayesianLinearModelAvro](https://github.com/linkedin/photon-ml/blob/master/photon-avro-schemas/src/main/avro/BayesianLinearModelAvro.avsc) Avro schema.

### Shaded Jar
[photon-all](https://github.com/linkedin/photon-ml/tree/master/photon-all) module releases a shaded jar containing all the required runtime dependencies of Photon ML, other than Spark and Hadoop. Shading is a robust way of creating fat/uber jars. It does not only package all dependencies into one single place, but also smartly renames a few selected class packages to avoid dependency conflicts. Although ```photon-all.jar``` is not a necessity, and it is fine for users to provide their own copies of dependences, it is highly recommended to be used in cluster environment where complex dependency conflicts could happen between system and user jars. (See [Gradle Shadow Plugin](https://github.com/johnrengelman/shadow) for more about shading).

Below is a command to build the photon-all jar:
```bash
./gradlew :photon-all:assemble
```

### Try It Out!

The easiest way to get started with Photon ML is to try the tutorial we created to demonstrate how GLMix models can be applied to build a personalized recommendation system. You can view the instructions on the wiki [here](https://github.com/linkedin/photon-ml/wiki/Photon-ML-Tutorial).

Alternatively, you can follow these steps to try Photon ML on your machine.

#### Install Spark

This step is platform-dependent. On OS X, you can install Spark with [Homebrew](http://brew.sh/) using the following command:

```
brew install apache-spark
```

For more information, see the [Spark docs](http://spark.apache.org/docs/latest/index.html).

#### Get and Build the Code

```
git clone git@github.com:linkedin/photon-ml.git
cd photon-ml
./gradlew build -x test -x integTest
```

#### Grab a Dataset

For this example, we'll use the "a1a" dataset, acquired from [https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/binary.html](https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/binary.html). Currently the Photon ML dataset converter supports only the LibSVM format.

```
curl -O https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/binary/a1a
curl -O https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/binary/a1a.t
```

Convert the data to the Avro format that the Photon ML drivers use.

```
mkdir -p a1a/train
mkdir -p a1a/test
pip install avro
python dev-scripts/libsvm_text_to_trainingexample_avro.py a1a dev-scripts/TrainingExample.avsc a1a/train/a1a.avro
python dev-scripts/libsvm_text_to_trainingexample_avro.py a1a.t dev-scripts/TrainingExample.avsc a1a/test/a1a.t.avro
```

The first command might be different, depending on the configuration of your system. If it fails, try your platform's standard approach for installing a Python library.

#### Train the Model

Now we're ready to train the model with Photon ML on your local dev box. Run the following command from the "photon-ml" directory:

```
spark-submit \
  --class com.linkedin.photon.ml.Driver \
  --master local[*] \
  --num-executors 4 \
  --driver-memory 1G \
  --executor-memory 1G \
  "./build/photon-all_2.10/libs/photon-all_2.10-1.0.0.jar" \
  --training-data-directory "./a1a/train/" \
  --validating-data-directory "./a1a/test/" \
  --format "TRAINING_EXAMPLE" \
  --output-directory "out" \
  --task "LOGISTIC_REGRESSION" \
  --num-iterations 50 \
  --regularization-weights "0.1,1,10,100" \
  --job-name "demo_photon_ml_logistic_regression"
```

Alternatively, to run the exact same training using the GAME training driver, use the following command:

```
spark-submit \
  --class com.linkedin.photon.ml.cli.game.GameTrainingDriver \
  --master local[*] \
  --num-executors 4 \
  --driver-memory 1G \
  --executor-memory 1G \
  "./build/photon-all_2.10/libs/photon-all_2.10-1.0.0.jar" \
  --input-data-directories "./a1a/train/" \
  --validation-data-directories "./a1a/test/" \
  --root-output-directory "out" \
  --feature-shard-configurations "name=globalShard,feature.bags=features" \
  --coordinate-configurations "name=global,feature.shard=globalShard,min.partitions=4,optimizer=LBFGS,tolerance=1.0E-6,max.iter=50,regularization=L2,reg.weights=0.1|1|10|100" \
  --coordinate-update-sequence "global" \
  --coordinate-descent-iterations 1 \
  --training-task "LOGISTIC_REGRESSION"
```

When this command finishes, you should have a new folder named "out" containing the trained model.

### Running Photon ML on Cluster Mode

In general, running Photon ML is no different from running other general Spark applications. As a result, using the ```spark-submit``` script in Spark’s ```bin``` directory we can run Photon ML on [different cluster modes](http://spark.apache.org/docs/latest/cluster-overview.html):
* [Spark Standalone Mode](http://spark.apache.org/docs/latest/spark-standalone.html)
* [Mesos](http://mesos.apache.org/)
* [YARN](http://hadoop.apache.org/docs/stable/hadoop-yarn/hadoop-yarn-site/YARN.html)

Below is a template for running a logistic regression training job with minimal setup on YARN. For running Photon ML using other cluster modes, the relevant arguments to ```spark-submit``` can be modified as detailed in [http://spark.apache.org/docs/latest/submitting-applications.html](http://spark.apache.org/docs/latest/submitting-applications.html).

```bash
spark-submit \
  --class com.linkedin.photon.ml.Driver \
  --master yarn \
  --deploy-mode cluster \
  --num-executors $NUM_EXECUTORS \
  --driver-memory $DRIVER_MEMORY \
  --executor-memory $EXECUTOR_MEMORY \
  "./build/photon-all_2.10/libs/photon-all_2.10-1.0.0.jar" \
  --training-data-directory "path/to/training/data" \
  --validating-data-directory "path/to/validating/data" \
  --output-directory "path/to/output/dir" \
  --task "LOGISTIC_REGRESSION" \
  --num-iterations 50 \
  --regularization-weights "0.1,1,10" \
  --job-name "demo_photon_ml_logistic_regression"
```

TODO: This example should be updated to use the GAME training driver instead.
There is also a more complex script demonstrating advanced options and customizations of using Photon ML at [example/run_photon_ml.driver.sh](https://github.com/linkedin/photon-ml/blob/master/examples/run_photon_ml_driver.sh).

Detailed usages are described via command:
```bash
./run_photon_ml.driver.sh [-h|--help]
```
**Note**: Not all configurations are currently exposed as options in the current script. Please directly modify the configurations if any customization is needed.

## Modules and directories
### Source code
- TODO: Photon ML modules are in need of a refactor. Once this is complete, this section will be updated.

### Other
- `build-scripts` contains scripts for Gradle tasks
- `buildSrc` contains Gradle plugin source code
- `dev-scripts` contains various scripts which may be useful for development
- `examples` contains a script which demonstrates how to run Photon ML from the command line
- `gradle` contains the Gradle wrapper jar
- `travis` contains scripts for controlling Travis CI test execution

## IntelliJ IDEA setup
When set up correctly, all the tests (unit and integration) can be run from IntelliJ IDEA, which is very helpful for development (IntelliJ IDEA's debugger can be used with all the tests). 
- Run `./gradlew idea`
- Open project as "New/Project from Existing Source", choose Gradle project, and set Gradle to use the local wrapper.

## How to Contribute
We welcome contributions. The following are good ways to get started: reporting an issue, fixing an existing issue, or participating in a discussion. For major functionality changes, it is highly recommended to exchange thoughts and designs with reviewers beforehand. Well communicated changes will have the highest probability of getting accepted.

## Reference
- XianXing Zhang, Yitong Zhou, Yiming Ma, Bee-Chung Chen, Liang Zhang and Deepak Agarwal. [GLMix: Generalized Linear Mixed Models For Large-Scale Response Prediction](http://www.kdd.org/kdd2016/papers/files/adf0562-zhangA.pdf). In 22nd SIGKDD Conference on Knowledge Discovery and Data Mining, 2016


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [19.0.1] - 2020-04-29
### Added
- Added publish to bintray functionality
### Changed
- Refactored feature normalization
- Refactored optimization states tracker
- Remove mean and variance type consistency check
### Fixed
- N/A

## [18.0.0] - 2020-03-06
### Added
- Added incremental training functionality
### Changed
- Refactored RDD join
- Refactored aggregators
### Fixed
- Fixed a bug to calculate RMSE off by sqrt(2)
- Fixed a bug which would keep L2 component of gradient and Hessian for intercept term
- Fixed a bug which would drop data which does not have initial random effect models
- Fixed a hyperparameter tuning dimension mismatch bug
- Fixed a bug which would filter out prior model coefficients for inactive features

## [17.0.0] - 2019-10-09
### Changed
- The intercept is no longer included in L2 regularization
### Deprecated
- Deprecated option to disable optimizer state tracking (it is now always enabled)
### Removed
- Removed unused/pointless modules
### Fixed
- Fixed bug which would treat data filtered by the 'active.data.lower.bound' as passive data

## [16.1.0] - 2019-06-17
### Added
- Added command line input for controlling minimum number of partitions for input data RDD to NameAndTermFeatureBagsDriver
### Fixed
- Fixed inefficient join in RandomEffectDataset construction which was causing excessive executor memory usage

## [16.0.0] - 2019-06-10
### Added
- Added a numerically stable function for computing Pearson correlation score to LocalDataset
### Changed
- Changed NameAndTermFeatureBagsDriver to read/write feature sets from/to text files using RDD functions
- Linear subspace projection is now the only kind of projection on random effect features and is now permanently enabled
### Fixed
- Warm start is now disabled for hyperparameter auto-tuning; this allows up to 10x the amount of iterations due to reduced size of RDD lineage

## [15.0.0] - 2019-05-15
### Changed
- Renamed 'updateModel' function in Coordinate class to 'trainModel'
### Removed
- Removed photon-diagnostics module
- Removed ModelTracker class
- Removed 'initializeModel' function from Coordinate class
### Fixed
- Fixed links in README
- Fixed bug which prevented training intercept-only models

## [14.0.0] - 2019-03-15
### Changed
- Improve performance of RandomEffectDataset construction
- Loosened permissions for Evaluator.evaluate function
### Removed
- Removed passive data lower bound as a tuning parameter
- Removed RDDLike.materializeOnce function
- Removed Coordinate.computeRegularizationTermValue function
### Fixed
- Fixed bug affecting reading Map types from AVRO using the AvroDataReader

## [13.0.0] - 2019-02-12
### Added
- Added option to compute approximate coefficient variances using the Hessian diagonal instead of the full Hessian matrix
### Changed
- Updated README
- Changed 'dataset' terminology to be consistent across library
- Refactored evaluation metric classes: created new EvaluationSuite class which manages a Set of Evaluators, the RDD they use, and the 'primary' Evaluator
- Replaced most references to 'mllib.linalg.Vectors' with 'ml.linalg.Vectors'
- Changed DataValidators to accept any numeric value, not just Double
- Map object loaded through AvroDataReader now drop any values that are non-String, non-numeric, or non-Boolean
### Removed
- Removed dependencies on photon-diagnostics module
### Fixed
- Fixed bug in MinHeapWithFixedCapacity which would clear the wrong heap
- Fixed various bugs in DataValidators
- Fixed bug which would ignore the L2 component of elastic net regularization

## [12.0.0] - 2018-09-11
### Added
### Changed
- BasicStatisticalSummary now stores index of intercept, if present
- Intercept column is ignored when writing BasicStatisticalSummary to disk
- Strengthened relationship between shifts and intercept in NormalizationContext
- Renamed DataSet and derived classes to Dataset and variants thereof
- Upgrade to Spark 2.3.0
- Disable Scala 2.10 crossbuild
### Removed
### Fixed
- Fixed inefficiency in NormalizationContext when neither shifts nor factors present

## [11.0.0] - 2018-08-27
### Added
- Added helper function to shrink search range when selecting hyperparamater vectors during Guassian auto-tuning
- Added hyperparameter tuner interface, for removing the direct dependency from Photon-ML on the hyperparameter tuning code
- Added Apache license check to Travis CI tests
- Added PhotonParams class, which contains shared parameter-management helper functions
### Changed
- Decouple hyperparameter tuning from the Photon-ML Evaluator family
- RandomEffectDataSetPartitioner now uses a single, lazy-evaluated backup partitioner instead of one generated per-instance requiring the backup partitioner
- Improved efficiency of MinHeapWithFixedCapacity merging algorithm
- Changed LBFGSB unit tests to be deterministic
- RandomEffectDataSetPartitioner now takes the active data upper bound into account, so that data is better distributed across partitions
### Removed
- Removed the 'NONE' training task
- Replaced Photon StorageLevel with Spark StorageLevel
- Removed unused feature filtering algorithms from LocalDataSet
### Fixed
- Fixed bug in using ExpectedImprovement for minimization problems

## [10.2.0] - 2018-07-31
### Added
- Enabled box constraint option for TRON in GLM driver
- Added evaluator for area under precision-recall curve (AUPR)
### Fixed
- Fixed bug in expected improvement

## [10.1.0] - 2018-07-17
### Added
- Added elastic net alpha to hyper-parameter auto-tuning vector
- Enabled Scala cross-building for versions 2.10.6 and 2.11.8
### Fixed
- Fixed bug which would unpersist models too early

## [10.0.0] - 2018-06-21
### Added

examples/run_photon_ml_driver.sh
#!/bin/bash

# Copyright 2017 LinkedIn Corp. All rights reserved.
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain a
# copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

# ###############################################################
#
# This is a script that demonstrates how to run photon ml from command line.
# It is only recommended to be used for dev purpose and please be aware a different user/app potentially need to adjust
# almost all the parameters to make the job running.
#
# ################################################################
DEFAULT_JOB_NAME="run-photon-ml-driver"
JOB_NAME=$DEFAULT_JOB_NAME
ADMIN_ACLS="$USER"

function usage() {
  echo "Usage: $0 [option...] hdfs_working_root" >&2
  echo
  echo "hdfs_working_root: The HDFS working root for the Photon driver. This script assumes the following structure: "
  echo "    train dataset input: [hdfs_working_root]/input/train"
  echo "    test dataset input: [hdfs_working_root]/input/test"
  echo "  And it will output towards: "
  echo "    model output: [hdfs_working_root]/results"
  echo "    model/feature summary: [hdfs_working_root]/summary"
  echo
  echo "Options:"
  echo "  -h, --help        Display the help message"
  echo "  -n, --job-name    Set the job name, default: $DEFAULT_JOB_NAME"
  echo "  -u, --admin-acls  The admin-acls of the Spark job, default is current user"
  echo "  -l, --jar-lib     The local directory that holds all jars needed for running the job. default: ./"
  echo
}

JAR_DIR="."
PHOTON_MAIN_JAR=""
LIBJARS=""

# Check if parameters options are given on the commandline
while :
do
    case "$1" in
      -h | --help)
          usage
          exit 0
          ;;
      -n | --job-name)
          JOB_NAME="$2"
          shift 2
          ;;
      -u | --admin-acls)
          ADMIN_ACLS="$2"
          shift 2
          ;;
      -l | --jar-lib)
          JAR_DIR="$2"
          shift 2
          ;;
      --) # End of all options
          shift
          break
          ;;
      -*)
          echo "Error: Unknown option: $1" >&2
          usage
          exit 1
          ;;
      *)  # No more options
          break
          ;;
    esac
done

WORKING_ROOT_DIR="$1"

REQUIRED_JARS=("joda-time-2.7.jar" "avro-1.7.5.jar" "avro-mapred-1.7.5.jar" "scopt_2.10-3.2.0.jar" "photon-schemas-0.0.2.jar" "xchart-2.5.1.jar" "batik-util-1.7.jar" "batik-awt-util-1.7.jar" "batik-svggen-1.7.jar" )
function resolveAllJars() {
  # Set main jar
  main_jar_count=`ls $JAR_DIR | grep photon-ml* | wc -l`
  if [[ $main_jar_count < 1 ]]
  then
    echo "Error: Cannot find photon-ml main jar in $JAR_DIR" >&2
    exit 1
  fi

  if [[ $main_jar_count > 1 ]]
  then
    echo "Error: Multiple photon-ml jars found in [$JAR_DIR]. Cannot decide which one to use" >&2
    exit 1
  fi

  PHOTON_MAIN_JAR="${JAR_DIR}/`ls $JAR_DIR | grep photon-ml*`"

  # Include all dependencies jar
  c=0
  for jar in ${REQUIRED_JARS[@]}
  do
    if [[ -f "$JAR_DIR/$jar" ]]
    then
      c=$((c+1))
    else
      echo "Error: expected dependency jar [$jar] not found in lib jar directory [$JAR_DIR]. Please download and put then in the jar lib." >&2
    fi
  done

  if [[ ${#REQUIRED_JARS[@]} != $c ]]
  then
    exit 1
  fi

  LIBJARS=$(printf ",${JAR_DIR}/%s" "${REQUIRED_JARS[@]}")
  LIBJARS=${LIBJARS:1}
}
resolveAllJars

TRAIN_ROOT="${WORKING_ROOT_DIR}/input"
TRAIN_INPUT_DIRS="${TRAIN_ROOT}/train"
VALIDATE_INPUT_DIRS="${TRAIN_ROOT}/test"
SUMMARY_OUTPUT_PATH="${WORKING_ROOT_DIR}/summary"
OUTPUT_DIR="${WORKING_ROOT_DIR}/results"

QUEUE="default"

EXECUTOR_MEMORY_MB=$((2 * 1024))
EXECUTOR_MEMORY="${EXECUTOR_MEMORY_MB}m" # amount of memory requested per container
EXECUTOR_MEMORY_OVERHEAD=$((EXECUTOR_MEMORY_MB / 4))

NUM_EXECUTORS=10 # number of container/executor requested for this application
MIN_PARTITIONS=$((NUM_EXECUTORS * 3))

DRIVER_MEMORY_MB=$((5 * 1024))
DRIVER_MEMORY="${DRIVER_MEMORY_MB}m" # amount of memory requested for driver
AKKA_FRAME_SIZE=256 # in MBs
KRYO_BUFFER_MAX="$((2 * 1024 - 1))m" # kryo.buffer.max must be < 2048m

MAX_RESULT_SIZE_MB=$((DRIVER_MEMORY_MB / 4))
MAX_RESULT_SIZE="${MAX_RESULT_SIZE_MB}m"

TASK_TYPE="LOGISTIC_REGRESSION"
INPUT_FORMAT="TRAINING_EXAMPLE"
NUM_ITERATIONS=20
LAMBDAS="1,10,50"
TOLERANCE="1e-5"
OPTMIZER_TYPE="TRON"
REGULARIZATION_TYPE="L2"
NORMALIZATION_TYPE="NONE"
VALIDATE_PER_ITERATION=true
USE_OPTIMIZATION_TRACKER=true
USE_INTERCEPT=true

# UNCOMMENT to use different options:
# ELASTIC_NET_ALPHA=0.1
# BOX_CONSTRAINTS=

if [ ! -z $ELASTIC_NET_ALPHA ]
then
  REGULARIZATION_TYPE_ARG="--regularization-type ELASTIC_NET"
  ELASTIC_NET_ARG="--elastic-net-alpha $ELASTIC_NET_ALPHA"
else
  REGULARIZATION_ARG="--regularization-type $REGULARIZATION_TYPE"
fi

if [ ! -z $BOX_CONSTRAINTS ]
then
  BOX_CONSTRAINTS_ARG="--coefficients-box-constraints $BOX_CONSTRAINTS"
fi

CMD_TO_RUN="
spark-submit \
--class com.linkedin.photon.ml.Driver \
--queue $QUEUE \
--master yarn-cluster \
--num-executors $NUM_EXECUTORS \
--driver-memory $DRIVER_MEMORY \
--executor-memory $EXECUTOR_MEMORY \
--conf \"spark.executor.extraJavaOptions=-XX:+PrintGCDetails -XX:+PrintGCTimeStamps\" \
--conf spark.ui.showConsoleProgress=false \
--conf spark.eventLog.enabled=true \
--conf spark.admin.acls=$ADMIN_ACLS \
--conf spark.driver.maxResultSize=$MAX_RESULT_SIZE \
--conf spark.shuffle.consolidateFiles=true \
--conf spark.yarn.executor.memoryOverhead=$EXECUTOR_MEMORY_OVERHEAD \
--conf spark.shuffle.io.preferDirectBufs=false \
--conf spark.akka.frameSize=$AKKA_FRAME_SIZE \
--conf spark.kryoserializer.buffer.max=$KRYO_BUFFER_MAX \
--jars $LIBJARS  \
$PHOTON_MAIN_JAR \
--job-name $JOB_NAME \
--task $TASK_TYPE \
--training-data-directory $TRAIN_INPUT_DIRS \
--output-directory $OUTPUT_DIR \
--validate-per-iteration $VALIDATE_PER_ITERATION \
--validating-data-directory $VALIDATE_INPUT_DIRS \
--optimization-tracker $USE_OPTIMIZATION_TRACKER \
--intercept $USE_INTERCEPT \
--convergence-tolerance $TOLERANCE \
--format $INPUT_FORMAT \
--optimizer $OPTMIZER_TYPE \
--normalization-type $NORMALIZATION_TYPE \
--num-iterations $NUM_ITERATIONS \
--summarization-output-dir $SUMMARY_OUTPUT_PATH \
--regularization-weights $LAMBDAS \
--min-partitions $MIN_PARTITIONS \
$REGULARIZATION_TYPE_ARG \
$ELASTIC_NET_ARG \
$BOX_CONSTRAINTS_ARG
"
echo "Execute command: $CMD_TO_RUN"

eval $CMD_TO_RUN
